Mathlib/Algebra/Module/Submodule/Pointwise.lean

```diff
@@ -277,4 +289,229 @@ scoped[Pointwise] attribute [instance] Submodule.pointwiseMulActionWithZero
 
 end
 
+section set_acting_on_submodules
+
+variable [AddCommMonoid M] [Module R M]
+
+/--
+Let `s ⊆ R` be a set and `N ≤ M` be a submodule, then `s • N` is the smallest submodule containing
+all `r • n` where `r ∈ s` and `n ∈ N`.
+-/
+protected def pointwiseSetSMul : SMul (Set R) (Submodule R M) where
```

Comment:
This generalizes to
```suggestion
protected def pointwiseSetSMul : SMul (Set S) (Submodule R M) where
```
right? How may of the other results also generalize for free?

Reply:
What is `S` here? Is it another ring of which `M` is a module?

Reply:
Everything upto induction principal should work mutatis mutandis. My hunch is that if we further assume `[SMulCommClass R S M]` then everything should work?

Reply:
So the generalized principal is this 
```lean
variable {s N} in
@[elab_as_elim]
lemma set_smul_inductionOn {motive : (x : M) → (_ : x ∈ s • N) → Prop}
    (x : M)
    (hx : x ∈ s • N)
    (smul₀ : ∀ ⦃r : S⦄ ⦃n : M⦄ (mem₁ : r ∈ s) (mem₂ : n ∈ N),
      motive (r • n) (mem_set_smul_of_mem_mem mem₁ mem₂))
    (smul₁ : ∀ (r : R) ⦃m : M⦄ (mem : m ∈ s • N) ,
      motive m mem → motive (r • m) (Submodule.smul_mem _ r mem)) --
    (add : ∀ ⦃m₁ m₂ : M⦄ (mem₁: m₁ ∈ s • N) (mem₂ : m₂ ∈ s • N),
      motive m₁ mem₁ → motive m₂ mem₂ → motive (m₁ + m₂) (Submodule.add_mem _ mem₁ mem₂))
    (zero : motive 0 (Submodule.zero_mem _)) :
    motive x hx :=
  let ⟨_, h⟩ := set_smul_le s N
    { carrier := { m | ∃ (mem : m ∈ s • N), motive m mem },
      zero_mem' := ⟨Submodule.zero_mem _, zero⟩
      add_mem' := fun ⟨mem, h⟩ ⟨mem', h'⟩ ↦ ⟨_, add mem mem' h h'⟩
      smul_mem' := fun r x ⟨mem, h⟩ ↦ ⟨_, smul₁ r mem h⟩ }
    (fun _ _ mem mem' ↦ ⟨mem_set_smul_of_mem_mem mem mem', smul₀ mem mem'⟩) hx
  h
```

But to change the next lemma 
```lean
lemma set_smul_eq_map [SMulCommClass R R N] [Module S N]  [SMulCommClass S R N] :
    s • N =
    Submodule.map
      (N.subtype.comp (Finsupp.lsum R <| DistribMulAction.toLinearMap _ _))
      (Finsupp.supported N R s) := by
```
I need `N` to be both a `R`-submodule and `S`-submodule, so I included `[Module S N]`, can I somehow say this module structure is restricted from `[Module S M]`

Reply:
Is perhaps [`Subbimodule`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Bimodule.html#Subbimodule.mk)

Reply:
I think I would just leave this one ungeneralized

Reply:
In [7d37123](https://github.com/leanprover-community/mathlib4/pull/7140/commits/7d371236ab626d2912422f17efab1875d1b68609) and [eecb383](https://github.com/leanprover-community/mathlib4/pull/7140/commits/eecb383e36ab97c536af331630659391d8d2df5b), the generalization has been done. Things that are not generalized were left some comments explaining why they can't or won't be generalized

Reply:
If we generalize so that `r • N` makes sense for all `r : S`, then `Submodule.singleton_set_smul`
  and `Submodule.singleton_set_smul` can be generalized as well. We can prove `x ∈ ({r} : Set S) • N ↔ ∃ (m : M), m ∈ N ∧ x = r • m` but we can't say `({r} : Set R) • N = r • N`